File: Theory of Algs/Java/PA2 - edited/TheoryofAlgsPA2/src/HuffmanEncoder.java


import java.util.Comparator;
import java.util.Hashtable;

class BinaryTreeNodeComparator implements Comparator<BinaryTreeNode> {

	@Override
	public int compare(BinaryTreeNode o1, BinaryTreeNode o2) {
		return o1.value - o2.value;
	}
}

public class HuffmanEncoder {

	private char alphabet[];
	private int frequencies[];
	private int sigma;

	private int encodingLength;
	private int tableSize;
	private Hashtable<Character, String> charToEncodingMapping; // hash table is instantiated

	public HuffmanEncoder(char alphabet[], int frequencies[], int sigma) {
		this.alphabet = alphabet;
		this.sigma = sigma;
		this.frequencies = frequencies;
		encodingLength = tableSize = 0;
		charToEncodingMapping = new Hashtable<>();
		encode();
	}

	private void encode() throws IndexOutOfBoundsException { // complete this method
		
		
	}

	private void buildTree() throws IndexOutOfBoundsException { // complete this method
		
	}

	private void createTable(BinaryTreeNode node, String encoding) { // complete this method
		charToEncodingMapping.put(node.character, encoding);
	}

	public String getEncoding(char c) {
		return charToEncodingMapping.get(c);
	}

	public int getSigma() {
		return sigma;
	}

	public int[] getFrequencies() {
		return frequencies;
	}

	public char[] getAlphabet() {
		return alphabet;
	}

	public int getTableSize() {
		return tableSize;
	}

	public int getEncodingLength() {
		return encodingLength;
	}
}

File: Theory of Algs/C#/PA2CS/C#/HuffmanEncoder.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace _PA2
{
    class BinaryTreeNodeComparator : IComparer<BinaryTreeNode>
    {
        public int Compare(BinaryTreeNode arg1, BinaryTreeNode arg2)
        {
            return arg1.value - arg2.value;
        }
    }

    public class HuffmanEncoder
    {

        private char[] alphabet;
        private int[] frequencies;
        private int sigma;
        private int encodingLength, tableSize;
        private Hashtable charToEncodingMapping;

        public HuffmanEncoder(char[] alphabet, int[] frequencies, int sigma)
        {
            this.alphabet = alphabet;
            this.sigma = sigma;
            this.frequencies = frequencies;
            encodingLength = tableSize = 0;
            charToEncodingMapping = new Hashtable();
            encode();
        }

        private void encode()
        {   // complete this method
            BinaryTreeNode root = buildTree();
            
            createTable(root, "");
            for (int i = 0; i < alphabet.Length; i++)
            {
                char c = alphabet[i];
                string str = getEncoding(c);
                encodingLength += frequencies[i] * str.Length;
                tableSize += str.Length + 8; // 8 bits for the character code
            }
        }

        private BinaryTreeNode buildTree()
        {   // complete this method
            PriorityQueue<BinaryTreeNode> PQ = new PriorityQueue<BinaryTreeNode>(new BinaryTreeNodeComparator());
            for (int i = 0; i < alphabet.Length; i++)
            {
                BinaryTreeNode x = new BinaryTreeNode(alphabet[i], frequencies[i]);
                PQ.add(x);
            }

            while (PQ.size() > 1)
            {
                BinaryTreeNode min = PQ.poll();
                BinaryTreeNode secMin = PQ.poll();
                BinaryTreeNode y = new BinaryTreeNode('\0', min.value + secMin.value);
                y.left = min;
                y.right = secMin;
                PQ.add(y);
                
            }
            return PQ.poll();
        }

        private void createTable(BinaryTreeNode node, String encoding)
        {   // complete this method
            if (node.left is null && node.right is null)
            {
                charToEncodingMapping.Add(node.character, encoding);
            }else
            {
                if (node.left is not null)
                    createTable(node.left, encoding + "0");
                if (node.right is not null)
                    createTable(node.right, encoding + "1");
            }
        }

        public string getEncoding(char c)
        {
            return charToEncodingMapping[c].ToString();
        }

        public int getSigma()
        {
            return sigma;
        }

        public int[] getFrequencies()
        {
            return frequencies;
        }

        public char[] getAlphabet()
        {
            return alphabet;
        }

        public int getTableSize()
        {
            return tableSize;
        }

        public int getEncodingLength()
        {
            return encodingLength;
        }
    }
}